server/src/main/java/alcatraz/server/Server.java

```diff
@@ -22,68 +19,69 @@ public Server(ReplicationInterface replication) throws RemoteException {
         this.state = replication.getSharedState();
     }
 
-    // TODO: instead of return boolean, throw an exception if player already exists
+    //MM20241121: implement appropriate exception objects
     @Override
-    public boolean registerPlayer(String playerName, ClientInterface client) throws RemoteException {
-
+    public void registerPlayer(String clientName, ClientInterface client) throws RemoteException {
         if (this.replication.isPrimary()) {
-            if (this.state.players.containsKey(playerName)) {
-                // TODO: throw an exception instead of returning false
-                return false;
+            if (this.state.players.containsKey(clientName)) {
+                //MM20241121: implement a dedicated exception class
+                throw new RemoteException();
             }
+
             // Store the Player with ClientInterface
-            this.state.players.put(playerName, new Player(client, playerName, "ip_address", "port"));
-            System.out.println(playerName + " registered");
-            return true;
+            //MM20241121: use a mutex at this stage!
+            this.state.players.put(clientName, new Player(client, clientName, "ip_address", "port"));
+            System.out.println(clientName + " registered");
+            this.replication.replicatePrimaryState();
             // TODO: there should be an early return instead of else
+            //MM20241121: why?
```

Comment:
It would demand less cognitive effort.
```java
if (!isNotPrimary) {
  forward();
  return;
}
// the rest function logic
```
Now we have to find the else every time that is 13 lines under the `if`, and then have to look even further if it is the last statement in the function. With early return you can see directly all the actions that happen if the server is not primary.

Reply:
totally agree! didn't know what you meant by early exit!